core/src/main/scala/kafka/network/RequestChannel.scala

```diff
@@ -245,9 +245,17 @@ class RequestChannel(val numProcessors: Int, val queueSize: Int) extends KafkaMe
   /** Send a response back to the socket server to be sent over the network */
   def sendResponse(response: RequestChannel.Response) {
     if (isTraceEnabled) {
-      val requestHeader = response.request.header
-      trace(s"Sending ${requestHeader.apiKey} response to client ${requestHeader.clientId} of " +
-        s"${response.responseSend.size} bytes.")
+      response.responseAction match {
+        case SendAction =>
+          val requestHeader = response.request.header
+          trace(s"Sending ${requestHeader.apiKey} response to client ${requestHeader.clientId} of ${response.responseSend.get.size} bytes.")
+        case NoOpAction =>
+          val requestHeader = response.request.header
+          trace(s"No need to send ${requestHeader.apiKey} response to client ${requestHeader.clientId}.")
+        case CloseConnectionAction =>
+          trace("The connection is being closed since request handler encountered an error.")
+        case _ => // it should not happen
```

Comment:
```
val message = response.responseAction match ...
trace(message)
```

Reply:
That would compute the message even if trace logging is not enabled.

Reply:
not clear how it possible, it with in `if (isTraceEnabled)` scope.

Reply:
My bad, I had missed the `if`. In that case, your suggestion sounds good.